In [1]:
import nltk
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

# Tokenization

In [3]:
from nltk.tokenize import sent_tokenize
text="""Dogs ,Canis lupus familiaris, are domesticated mammals,
not natural wild animals. They were originally bred from
wolves. They have been bred by humans for a long time, and
were the first animals ever to be domesticated.
Today, some dogs are used as pets, others are used to help
humans do their work. They are a popular pet because they
are usually playful, friendly, loyal and listen to humans.
Thirty million dogs in the United States are registered as
pets. Dogs eat both meat and vegetables, often mixed
together and sold in stores as dog food. Dogs often have
jobs, including as police dogs, army dogs, assistance dogs,
fire dogs, messenger dogs, hunting dogs, herding dogs, or
rescue dogs.
They are sometimes called "canines" from the Latin word for dog
- canis. Sometimes people also use "dog" to describe other
canids, such as wolves. A baby dog is called a pup or puppy.
A dog is called a puppy until it is about one year old.
Dogs are sometimes referred to as "man’s best friend" because
they are kept as domestic pets and are usually loyal and
like being around humans."""
tokenized_text=sent_tokenize(text)
print(tokenized_text)

['Dogs ,Canis lupus familiaris, are domesticated mammals,\nnot natural wild animals.', 'They were originally bred from\nwolves.', 'They have been bred by humans for a long time, and\nwere the first animals ever to be domesticated.', 'Today, some dogs are used as pets, others are used to help\nhumans do their work.', 'They are a popular pet because they\nare usually playful, friendly, loyal and listen to humans.', 'Thirty million dogs in the United States are registered as\npets.', 'Dogs eat both meat and vegetables, often mixed\ntogether and sold in stores as dog food.', 'Dogs often have\njobs, including as police dogs, army dogs, assistance dogs,\nfire dogs, messenger dogs, hunting dogs, herding dogs, or\nrescue dogs.', 'They are sometimes called "canines" from the Latin word for dog\n- canis.', 'Sometimes people also use "dog" to describe other\ncanids, such as wolves.', 'A baby dog is called a pup or puppy.', 'A dog is called a puppy until it is about one year old.', 'Dogs are somet

In [4]:
print(len(tokenized_text))

13


In [5]:
print(tokenized_text[2])

They have been bred by humans for a long time, and
were the first animals ever to be domesticated.


Q: How many sentences in the paragraph?
<br>
A: 13 

Q: What is the third sentences in the paragraph?<br>
A: They have been bred by humans for a long time, and were the first animals ever to be domesticated.

Word Tokenization

In [7]:
from nltk.tokenize import word_tokenize
tokenized_word=word_tokenize(text)
print(tokenized_word)

['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'are', 'domesticated', 'mammals', ',', 'not', 'natural', 'wild', 'animals', '.', 'They', 'were', 'originally', 'bred', 'from', 'wolves', '.', 'They', 'have', 'been', 'bred', 'by', 'humans', 'for', 'a', 'long', 'time', ',', 'and', 'were', 'the', 'first', 'animals', 'ever', 'to', 'be', 'domesticated', '.', 'Today', ',', 'some', 'dogs', 'are', 'used', 'as', 'pets', ',', 'others', 'are', 'used', 'to', 'help', 'humans', 'do', 'their', 'work', '.', 'They', 'are', 'a', 'popular', 'pet', 'because', 'they', 'are', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'and', 'listen', 'to', 'humans', '.', 'Thirty', 'million', 'dogs', 'in', 'the', 'United', 'States', 'are', 'registered', 'as', 'pets', '.', 'Dogs', 'eat', 'both', 'meat', 'and', 'vegetables', ',', 'often', 'mixed', 'together', 'and', 'sold', 'in', 'stores', 'as', 'dog', 'food', '.', 'Dogs', 'often', 'have', 'jobs', ',', 'including', 'as', 'police', 'dogs', ',', 'army', 'dogs', ',', 

In [8]:
print(len(tokenized_word))

229


In [9]:
print(len(word_tokenize(tokenized_text[0])))

15


In [10]:
print(len(word_tokenize(tokenized_text[12])))

30


Q: How many words in the paragraph?<br>
A: 229

Q: How many words in the first sentence?<br>
A: 15

Q: How many words in the last sentence?<br>
A: 30

Q: Do you find repeat words?<br>
A: Yes

# Frequency Distribution

In [13]:
from nltk.probability import FreqDist
fdist = FreqDist(tokenized_word)
print(fdist)

<FreqDist with 120 samples and 229 outcomes>


Q: What is the output from print(fdist)?<br>
A: FreqDist with 120 samples and 229 outcomes

In [16]:
fdist

FreqDist({',': 18, '.': 13, 'are': 10, 'dogs': 10, 'as': 7, 'and': 6, 'to': 5, 'dog': 5, 'Dogs': 4, 'They': 4, ...})

Q: How about the most common words?<br>
A: ',' 18 times

In [17]:
fdist.most_common(10)

[(',', 18),
 ('.', 13),
 ('are', 10),
 ('dogs', 10),
 ('as', 7),
 ('and', 6),
 ('to', 5),
 ('dog', 5),
 ('Dogs', 4),
 ('They', 4)]

# Stopwords

In [18]:
my_stopwords = ["is", "a", "this", ","]

from nltk.corpus import stopwords
nltk_stop_words=set(stopwords.words("english"))
print(nltk_stop_words)

{'himself', 'down', 'herself', 'doesn', 'all', 'few', "couldn't", 'itself', 'and', 'against', 'ain', 'his', 'she', 'him', 'too', 'their', 'didn', 'so', 'themselves', "hadn't", 'its', 'we', 'when', 'has', 'can', "needn't", 'that', 'isn', 'whom', 'mightn', 'where', 'yours', 'he', "aren't", "didn't", 'most', 'm', 'weren', 'y', 'with', 'not', 'hadn', 'at', 'our', 'out', 'only', 'off', 'shan', 'it', 'ours', 'had', 'hers', "hasn't", 'having', 'to', 'how', 'this', 'very', 'the', 'of', 'there', 'haven', 'theirs', 'couldn', 'shouldn', 'be', 's', 'wasn', 'same', 'why', 'needn', "mightn't", 'your', "won't", 'yourselves', 'am', 'does', 'mustn', "don't", 'yourself', 'then', 'who', 'up', 'more', 're', 'do', 'before', 'other', 'what', "mustn't", 'was', 'but', 'were', 'own', 'while', 'myself', 'these', 'both', 'will', "you'll", 'until', "she's", 'they', 'or', 'll', 'a', 'under', 'further', 'is', 'about', 'them', 'above', 'again', "that'll", 'should', "shouldn't", 'on', 'aren', 'hasn', 'just', 'doing',

# Removing Stopwords

In [19]:
filtered_words=[]
for w in tokenized_word:
    if w not in my_stopwords:
        filtered_words.append(w)
print("Tokenized Sentence:",tokenized_word)
print("\n")
print("Filterd Sentence:",filtered_words)

Tokenized Sentence: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'are', 'domesticated', 'mammals', ',', 'not', 'natural', 'wild', 'animals', '.', 'They', 'were', 'originally', 'bred', 'from', 'wolves', '.', 'They', 'have', 'been', 'bred', 'by', 'humans', 'for', 'a', 'long', 'time', ',', 'and', 'were', 'the', 'first', 'animals', 'ever', 'to', 'be', 'domesticated', '.', 'Today', ',', 'some', 'dogs', 'are', 'used', 'as', 'pets', ',', 'others', 'are', 'used', 'to', 'help', 'humans', 'do', 'their', 'work', '.', 'They', 'are', 'a', 'popular', 'pet', 'because', 'they', 'are', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'and', 'listen', 'to', 'humans', '.', 'Thirty', 'million', 'dogs', 'in', 'the', 'United', 'States', 'are', 'registered', 'as', 'pets', '.', 'Dogs', 'eat', 'both', 'meat', 'and', 'vegetables', ',', 'often', 'mixed', 'together', 'and', 'sold', 'in', 'stores', 'as', 'dog', 'food', '.', 'Dogs', 'often', 'have', 'jobs', ',', 'including', 'as', 'police', 'dogs', ',', '

Q: Can you apply the stopwords provided by NLTK (nltk stop words) to your text?

In [20]:
filtered_words=[]
for w in tokenized_word:
    if w not in nltk_stop_words:
        filtered_words.append(w)
print("Tokenized Sentence:",tokenized_word)
print("\n")
print("Filterd Sentence:",filtered_words)

Tokenized Sentence: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'are', 'domesticated', 'mammals', ',', 'not', 'natural', 'wild', 'animals', '.', 'They', 'were', 'originally', 'bred', 'from', 'wolves', '.', 'They', 'have', 'been', 'bred', 'by', 'humans', 'for', 'a', 'long', 'time', ',', 'and', 'were', 'the', 'first', 'animals', 'ever', 'to', 'be', 'domesticated', '.', 'Today', ',', 'some', 'dogs', 'are', 'used', 'as', 'pets', ',', 'others', 'are', 'used', 'to', 'help', 'humans', 'do', 'their', 'work', '.', 'They', 'are', 'a', 'popular', 'pet', 'because', 'they', 'are', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'and', 'listen', 'to', 'humans', '.', 'Thirty', 'million', 'dogs', 'in', 'the', 'United', 'States', 'are', 'registered', 'as', 'pets', '.', 'Dogs', 'eat', 'both', 'meat', 'and', 'vegetables', ',', 'often', 'mixed', 'together', 'and', 'sold', 'in', 'stores', 'as', 'dog', 'food', '.', 'Dogs', 'often', 'have', 'jobs', ',', 'including', 'as', 'police', 'dogs', ',', '

# Lexicon Normalization

Q: What is the output?<br>
A: connect

In [22]:
from nltk.stem import PorterStemmer
stem = PorterStemmer()
word = "connection"
stem.stem(word)

'connect'

Q: Can you try some other words such as derivational, multiplying?<br>
A: deriv , multipli

In [23]:
word1 = "derivational"
stem.stem(word1)

'deriv'

In [24]:
word2 = "multiplying"
stem.stem(word2)

'multipli'

# Lemmatization

Q: What is the output?

In [25]:
from nltk.stem import PorterStemmer
stem = PorterStemmer()
stemmed_words=[]
for w in filtered_words:
    stemmed_words.append(stem.stem(w))
print("Filtered Words:",filtered_words)
print("\n")
print("Stemmed Words:",stemmed_words)

Filtered Words: ['Dogs', ',', 'Canis', 'lupus', 'familiaris', ',', 'domesticated', 'mammals', ',', 'natural', 'wild', 'animals', '.', 'They', 'originally', 'bred', 'wolves', '.', 'They', 'bred', 'humans', 'long', 'time', ',', 'first', 'animals', 'ever', 'domesticated', '.', 'Today', ',', 'dogs', 'used', 'pets', ',', 'others', 'used', 'help', 'humans', 'work', '.', 'They', 'popular', 'pet', 'usually', 'playful', ',', 'friendly', ',', 'loyal', 'listen', 'humans', '.', 'Thirty', 'million', 'dogs', 'United', 'States', 'registered', 'pets', '.', 'Dogs', 'eat', 'meat', 'vegetables', ',', 'often', 'mixed', 'together', 'sold', 'stores', 'dog', 'food', '.', 'Dogs', 'often', 'jobs', ',', 'including', 'police', 'dogs', ',', 'army', 'dogs', ',', 'assistance', 'dogs', ',', 'fire', 'dogs', ',', 'messenger', 'dogs', ',', 'hunting', 'dogs', ',', 'herding', 'dogs', ',', 'rescue', 'dogs', '.', 'They', 'sometimes', 'called', '``', 'canines', "''", 'Latin', 'word', 'dog', '-', 'canis', '.', 'Sometimes', '

In [26]:
from nltk.stem.wordnet import WordNetLemmatizer
lem = WordNetLemmatizer()
word = "connecting"
lem.lemmatize(word, pos="v")

'connect'

Note that pos is a part of speech parameter. Let’s change pos to see the result.

In [27]:
from nltk.stem.wordnet import WordNetLemmatizer
lem = WordNetLemmatizer()
word = "flying"
print(lem.lemmatize(word, pos="v"))
print(lem.lemmatize(word, pos="n"))

fly
flying


Q: What is the different between using v and n?<br>
A: v = verb and n = noun

# POS Tagging

In [28]:
sent = "I am studying at ICT"
tokens=nltk.word_tokenize(sent)
print(tokens)
nltk.pos_tag(tokens)

['I', 'am', 'studying', 'at', 'ICT']


[('I', 'PRP'),
 ('am', 'VBP'),
 ('studying', 'VBG'),
 ('at', 'IN'),
 ('ICT', 'NNP')]

Q: Can you try to interpret the result?<br>
A: PRP = Personal pronoun<br>
   VBP = Verb, non-3rd person singular present<br>
   VBG = Verb, gerund or present participle<br> 
   IN = Preposition or subordinating conjunction<br> 
   NNP = Proper noun, singular<br>